official-antistasi-community/A3-Antistasi
# General parameter saving with individual overrides

### What type of PR is this.
1. [ ] Bug
2. [ ] Change
3. [X] Enhancement

### What have you changed and why?
I never liked the parameter saving system in KP Liberation or CO10 Escape because you have to set every non-default parameter each time you change one, which is annoying and extremely error-prone. So I wrote this instead.

The principle is that every parameter defaults to a string named "Default (some value)". If this parameter is unchanged, it uses the value from the save game if there is one, or the default otherwise. If the parameter is changed, it overrides any other source. So you can simply change the params you want to change, and the rest will remain as they were before.

The init order was changed slightly to load saved params before initVar is executed, as initVar depends on some params (eg. gameMode, enabled DLC). Clients (including HCs) are blocked from running initVar until the server has finished loading and publishing params. Shouldn't have any significant impact on loading times.

I also moved the global variables that are currently user-adjustable (spawn distance, civ count, max units) to the same system. This is a bit questionable as it doesn't save much complexity at the moment, but it should simplify things later if we want to expose more adjustable parameters to the in-game UI.

Otherwise, behaviour should be unchanged. Parameter names, variable names and defaults are untouched. Variables that were public or bool-converted before retain those qualities. Inconsistencies, dead parameters and weird names are retained. I could do a cleanup here but I didn't want to confuse the issues.

### Please specify which Issue this PR Resolves.
closes #650

### Please verify the following and ensure all checks are completed.
1. [X] Have you loaded the mission in singleplayer?
2. [X] Have you loaded the mission in LAN host?
3. [X] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [X] Yes (Please provide further detail below.)

It's not based on bugfix and so will conflict heavily with that branch without modifications. I don't know what milestone this should go in, so I'm leaving it on unstable at the moment.

There's some spammy debug code in initParams that should be disabled before merging.

### How can the changes be tested?
Fiddle with parameters, start games, check the outcome in the RPT debug spam. Try with new games, old saves and new saves.
